### PR TITLE
feat: add trace_replayBlockTransactions

### LINKED ISSUE #2010: Implement `trace_replayTransaction` + `trace_replayBlockTransactions``
### Describe the feature

ref #1225 

The `trace_replayTransaction` is unimplemented:

https://github.com/paradigmxyz/reth/blob/8987f4e8be60b92d596acc21a2b27ecc4ede9f1a/crates/rpc/rpc/src/trace.rs#L227-L228

This can be implemented together with `trace_replayBlockTransactions` because for `trace_replayTransaction` we need to replay the block until the target transaction anyways.

## TODO
* fetch block + block's `env`
* execute all transactions (with inspector or without)

I believe some parts can be reused after some modifications:

https://github.com/paradigmxyz/reth/blob/8987f4e8be60b92d596acc21a2b27ecc4ede9f1a/crates/rpc/rpc/src/trace.rs#L71-L81

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chirag-bgh**
Closes #2010 

Implements `trace_replayBlockTransactions`

### CODE REVIEW

File: crates/rpc/rpc/src/trace.rs [hunk 1]
```diff
@@ -164,6 +164,38 @@ where
             .ok_or_else(|| EthApiError::TransactionNotFound)?
     }
 
+    /// Replays all transaction in a block
+    pub async fn replay_block_transactions(
+        &self,
+        block_id: BlockId,
+        trace_types: HashSet<TraceType>,
+    ) -> Result<Option<Vec<TraceResultsWithTransactionHash>>> {
```
Comment (mattsse) [inline crates/rpc/rpc/src/trace.rs:172]: this is basically the same thing as `trace_block`, but the recorded traces differ: 

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/trace.rs#L238-L241

what we should do here is:

Convert `trace_block` into a helper function that can also unify `replay_block_transactions`. This new function should accept a `TracingInspectorConfig` and a callback function `FnOnce()` like this:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/api/transactions.rs#L136-L140

so both trace block functions call this new helper functions but with a different config and callback (that converts the inspector into traces)

does this make sense?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/rpc/rpc/src/trace.rs (hunk 1 above)
Comment (chirag-bgh) [reply crates/rpc/rpc/src/trace.rs:172]: Okay 
Yes this make sense

**Comment by mattsse** [comment]
hey @chirag-bgh 
I'd like to get this PR over the line, 
Do you have enough bandwith to finish, or do you need pointers on the requested changes?

**Comment by gakonst** [comment]
Bump @chirag-bgh - we're sprinting towards release so would love to merge this asap. If you cannot prioritize it's all good! We can take it over, thank you for getting the ball rolling..

**Comment by chirag-bgh** [comment]
> Bump @chirag-bgh - we're sprinting towards release so would love to merge this asap. If you cannot prioritize it's all good! We can take it over, thank you for getting the ball rolling..

I'm very sorry for keeping this stale. You can take it over!

**Review by mattsse — APPROVED** [review]

**Pull request merged**